Many small usability enhancements to fully-saturated flow in PorousFlow

WilkAndy opened the pull request:
This commit contains code, doco and tests for a number of related (but small) enhancements to PorousFlow, all centering around the PorousFlowFullySaturated Action.

1. PorousFlow didn't have any specialised Kernels that did full upwinding for fully-saturated single-phase flow.  Instead, users had to employ the general multi-phase case with trivial relative permeability, which was computationally inefficient.  Therefore:
   - PorousFlowFullySaturatedAdvectiveFlux Kernel added: multi-component Darcy flow for single-phase, fully-saturated flow, and uses full upwinding
   - PorousFlowFullySaturatedUpwindHeatAdvection Kernel added: heat advection by a fully-saturated, single-phase fluid, and uses full upwinding.
   - PorousFlowFullySaturatedAction uses these Kernels when stabilization=Full (previously, the unstabilized Kernels were used even if the user desired full stabilization, ooops)


2. Occasionally, it is numerically efficient to not multiply the fluid-flow equations by fluid density.  However, the PorousFlowFullySaturated Action did not offer this functionality.  Therefore:
   - PorousFlowMassTimeDerivative now has a multiply_by_density flag
   - PorousFlowMassVolumetricExpansion now has a multiply_by_density flag
   - PorousFlowFullySaturated action now passes multiply_by_density to all fluid-flow related objects correctly


3. In certain situations, the nodal strain Material values should be stored at the nearest quadpoint to the nodes.  However, this could not be activated via the PorousFlow Actions, so:
   - strain_at_nearest_qp added to PorousFlowActionBase and propagated through to relevant Kernels


4. In some models, it is useful to save the nodal fluid-component rates coming from the time-derivative Kernel, and this is now possible to do via a PorousFlow Action:
   - PorousFlowSinglePhaseBase now allows save_component_rate_in to be specified

5. The fluid properties (PorousFlowBrine and PorousFlowSingleComponentFluid) has been refactored in PorousFlowSinglePhaseBase, in order to make the interface more consistent (in my opinion) between these two, and:
   - for Brine, allow the NaCl variable to be a NonlinearVariable or an AuxVariable
   - allow user to specify their own Material that will calculate fluid density, viscosity, etc

6. The PorousFlowSinglePhase Actions can now set temperature_unit, pressure_unit and time_unit
    
Refs #16841

@cpgr - do you have time to review this (after i've squashed commits, which will take an hour or so)
@rpodgorney - this is the first contribution in our new collaboration

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/682382/) on 10eec80 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16995/site)

This comment will be updated on new commits.

WilkAndy committed:
This commit contains code for a number of related (but small) enhancements to PorousFlow, all centering around the PorousFlowFullySaturated Action.

1. PorousFlow didn't have any specialised Kernels that did full upwinding for fully-saturated single-phase flow.  Instead, users had to employ the general multi-phase case with trivial relative permeability, which was computationally inefficient.  Therefore:
   - PorousFlowFullySaturatedAdvectiveFlux Kernel added: multi-component Darcy flow for single-phase, fully-saturated flow, and uses full upwinding
   - PorousFlowFullySaturatedUpwindHeatAdvection Kernel added: heat advection by a fully-saturated, single-phase fluid, and uses full upwinding.
   - PorousFlowFullySaturatedAction uses these Kernels when stabilization=Full (previously, the unstabilized Kernels were used even if the user desired full stabilization, ooops)

2. Occasionally, it is numerically efficient to not multiply the fluid-flow equations by fluid density.  However, the PorousFlowFullySaturated Action did not offer this functionality.  Therefore:
   - PorousFlowMassTimeDerivative now has a multiply_by_density flag
   - PorousFlowMassVolumetricExpansion now has a multiply_by_density flag
   - PorousFlowFullySaturated action now passes multiply_by_density to all fluid-flow related objects correctly

3. In certain situations, the nodal strain Material values should be stored at the nearest quadpoint to the nodes.  However, this could not be activated via the PorousFlow Actions, so:
   - strain_at_nearest_qp added to PorousFlowActionBase and propagated through to relevant Kernels

4. In some models, it is useful to save the nodal fluid-component rates coming from the time-derivative Kernel, and this is now possible to do via a PorousFlow Action:
   - PorousFlowSinglePhaseBase now allows save_component_rate_in to be specified

5. The fluid properties (PorousFlowBrine and PorousFlowSingleComponentFluid) has been refactored in PorousFlowSinglePhaseBase, in order to make the interface more consistent (in my opinion) between these two, and:
   - for Brine, allow the NaCl variable to be a NonlinearVariable or an AuxVariable
   - allow user to specify their own Material that will calculate fluid density, viscosity, etc

6. The PorousFlowSinglePhase Actions can now set temperature_unit, pressure_unit and time_unit

Refs #16841

WilkAndy committed:
Doco.  Refs #16841

WilkAndy committed:
Tests.  Fixes #16841

cpgr commented:
I'll take a look today

WilkAndy commented:
> I'll take a look today

The code is not huge.  The large number of lines comes from all the tests.

cpgr commented on modules/combined/examples/geochem-porous_flow/forge/porous_flow.i:
```diff
@@ -630,6 +475,11 @@
   [../]
+  [./relperm] # this is necessary because of the use_mobility = true used in the PorousFlowPeacemanBorehole objects
+    type = PorousFlowRelativePermeabilityConst
+    at_nodes = true
```
Is the `at_nodes` required? If it is, remind me to fix this so that it isn't

WilkAndy replied:
Yea, `at_nodes` is needed, otherwise a 

> ```
> [blah]
>   blah
>   blah
> []
> ```
> 
> 
> while others have
> 
> ```
> [./blah]
>   blah
>   blah
> [../]
> ```

I just submitted a new issue on this (#17024) - i think it's more widespread than the test/example files in this PR

WilkAndy replied:
?  The above comment about #17024 was nothing to do with the `at_nodes` remark, sorry.

The `at_nodes` param is needed, otherwise both the Action and the AddJoiner will create the `at_nodes = False` version.   Hope that makes sense!

WilkAndy replied:
I just submitted an issue on this `at_nodes` thing - #17025

cpgr commented on modules/porous_flow/doc/content/source/actions/PorousFlowBasicTHM.md:
```diff
@@ -1,29 +1,22 @@
-are somewhat inefficient in the fully-saturated case since Material
-Properties such as relative permeabilities and saturations actually do
-not need to be computed.
+In this equation, the fluid density, $\rho$, appears in parenthases, because the user has the option of including it or not using the `multiply_by_density` flag.  Note that the fluid-mass time derivative is close to linear, and is perfectly linear if `multiply_by_density=false`, and this also almost linearises the flow term.  Extremely good nonlinear convergence should therefore be expected, but  ...
```
typo: parentheses

WilkAndy replied:
done

cpgr commented on modules/porous_flow/src/actions/PorousFlowUnsaturated.C:
```diff
@@ -75,6 +75,8 @@ PorousFlowUnsaturated::PorousFlowUnsaturated(const InputParameters & params)
     _capillary_pressure_name("PorousFlowUnsaturated_CapillaryPressureVG")
 {
+  if (_stabilization == StabilizationEnum::None)
+    paramError("stabilization", "PorousFlowUnsaturated cannot use no stabilization");
```
Deliberate double negative? 🤣

WilkAndy replied:
oh, dreadful, sorry.  Changed.

cpgr commented on modules/porous_flow/src/kernels/PorousFlowFullySaturatedAdvectiveFlux.C:
```diff
@@ -0,0 +1,75 @@
+               _fluid_viscosity[nodenum][phase] -
+           _mass_fractions[nodenum][phase][_fluid_component] *
+               _dfluid_viscosity_dvar[nodenum][phase][pvar] /
+               std::pow(_fluid_viscosity[nodenum][phase], 2);
```
`Utility::pow<2>()` instead?

WilkAndy replied:
yep, done

cpgr commented on modules/porous_flow/src/kernels/PorousFlowFullySaturatedAdvectiveFlux.C:
```diff
@@ -0,0 +1,75 @@
+        _dfluid_density_node_dvar[nodenum][phase][pvar] / _fluid_viscosity[nodenum][phase];
+  dm -= _mass_fractions[nodenum][phase][_fluid_component] * _fluid_density_node[nodenum][phase] *
+        _dfluid_viscosity_dvar[nodenum][phase][pvar] /
+        std::pow(_fluid_viscosity[nodenum][phase], 2);
```
same?

WilkAndy replied:
yep, done

cpgr commented on modules/porous_flow/src/kernels/PorousFlowFullySaturatedUpwindHeatAdvection.C:
```diff
@@ -0,0 +1,55 @@
+        _fluid_viscosity[nodenum][phase];
+  dm -= _enthalpy[nodenum][phase] * _fluid_density_node[nodenum][phase] *
+        _dfluid_viscosity_dvar[nodenum][phase][pvar] /
+        std::pow(_fluid_viscosity[nodenum][phase], 2);
```
same

WilkAndy replied:
yep, done

cpgr commented on modules/porous_flow/src/kernels/PorousFlowMassVolumetricExpansion.C:
```diff
@@ -71,22 +85,26 @@ PorousFlowMassVolumetricExpansion::PorousFlowMassVolumetricExpansion(
               "PorousFlowMassVolumetricExpansion: Size of fluid density = "
                   << num_phases << " size of fluid saturation = " << _fluid_saturation[_i].size()
                   << " but both these must be equal to the number of phases in the system");
+  */
```
Did you want this commented code left in?

WilkAndy replied:
no, i left it there by accident.  i've removed it now

cpgr commented on modules/porous_flow/test/tests/actions/fullsat_brine.i:
```diff
@@ -27,9 +27,10 @@
-  use_brine = true
-  nacl_index = 0
+  fluid_properties_type = PorousFlowBrine
+  nacl_name = nacl
```
I like this syntax much better!

WilkAndy replied:
good :-)

cpgr commented on modules/porous_flow/test/tests/fluids/simple_fluid_yr_MPa_C_action.i:
```diff
@@ -0,0 +1,114 @@
+  [../]
+  [./density]
+    type = ElementIntegralMaterialProperty
+    mat_prop = 'PorousFlow_fluid_phase_density_qp0'
```
Maybe we should make a porous flow version of this in the future where the users don't have to know the internal names of our material properties. Not for this PR though.

WilkAndy replied:
I don't think this is necessary, because most users will use `PorousFlowPropertyAux` to populate an AuxVariable, and then use standard MOOSE postprocessors.  This file uses this obscure stuff just because i (or you - can't remember who did this input file originally) wanted to short-circuit that process to keep the input file small.

cpgr replied:
Good point. We probably did this first version before we had the aux kernels

cpgr reviewed (commented):
Might be minor and a pain to fix, but some of the input files have 
```
[blah]
  blah
  blah
[]
```
while others have
```
[./blah]
  blah
  blah
[../]
```

and some have a mixture.

cpgr commented:
It would be nice for usability to treat the brine the same as the other fluids, but I'm not sure how to do that any better than you have here.

WilkAndy commented:
> It would be nice for usability to treat the brine the same as the other fluids, but I'm not sure how to do that any better than you have here.

Yea, i agree, but i think that would require a refactor of PorousFlowBrine and/or PorousFlowSingleComponentFluid.  I'm not sure the pain of that offsets the gain.  What do you think?

WilkAndy committed:
@cpgr changes

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/684758/) on 8ab7b76 : invalidated by @WilkAndy

moose threads test: materials/stateful_prop.adaptivity/spatially_varying failed

WilkAndy commented:
I think this is ready for merging, @loganharbour

WilkAndy commented:
I think this is ready for merging, @loganharbour
